runtime/alloc.c

```diff
@@ -83,70 +76,80 @@ Caml_inline value do_alloc_small(mlsize_t wosize, tag_t tag, value* vals)
     CAMLdrop;                                               \
   } while (0)
 
-  Alloc_small(v, wosize, tag, Enter_gc);
-  for (i = 0; i < wosize; i++) {
-    Field(v, i) = vals[i];
-  }
-  return v;
+/* This has to be done with a macro, rather than an inline function, since
+   otherwise the wosize parameter to CAMLlocalN expands to be a VLA, which
+   breaks MSVC. */
+#define Do_alloc_small(wosize, tag)                     \
+{                                                       \
+  Caml_check_caml_state();                              \
+  value v;                                              \
+  mlsize_t i;                                           \
+  CAMLassert ((tag) < 256);                             \
+                                                        \
+  Alloc_small(v, wosize, tag, Enter_gc_preserve_vals);  \
+  for (i = 0; i < (wosize); i++) {                      \
+    Field(v, i) = vals[i];                              \
+  }                                                     \
+  return v;                                             \
 }
 
 CAMLexport value caml_alloc_1 (tag_t tag, value a)
 {
-  value v[1] = {a};
-  return do_alloc_small(1, tag, v);
+  value vals[1] = {a};
+  Do_alloc_small(1, tag);
 }
 
 CAMLexport value caml_alloc_2 (tag_t tag, value a, value b)
 {
-  value v[2] = {a, b};
-  return do_alloc_small(2, tag, v);
+  value vals[2] = {a, b};
+  Do_alloc_small(2, tag);
 }
 
 CAMLexport value caml_alloc_3 (tag_t tag, value a, value b, value c)
 {
-  value v[3] = {a, b, c};
-  return do_alloc_small(3, tag, v);
+  value vals[3] = {a, b, c};
+  Do_alloc_small(3, tag);
```

Comment:
Actually, could this be just

```c
CAMLexport value caml_alloc_3 (tag_t tag, value a, value b, value c)
{
  Do_alloc_small(3, tag, {a, b, c});
}
```

?

With this form we can bind `vals` in `Do_alloc_small` itself, and thus reduce the scope of the code that has to reason about `vals` as a magical/fixed name.

Reply:
I don't think it can work - the `{ ... }` is an initialiser syntax and it doesn't work as a bracket for the macro (that gets seen by the preprocessor as a 5-argument macro call)

Reply:
Just add parentheses!  `Do_alloc_small(3, tag, ({a, b, c}))` should pass the preprocessor.  But then you could just omit the braces and add them back in the macro.

Reply:
I'd tried that - but then you end up with `value vals[3] = ({a, b, c})` which is a syntax error. I hadn't considered passing them as `(a, b, c)` and then restoring the braces in the macro, but that doesn't appear to work either ("left-hand operand of comma expression has no effect" presumably arising from `value vals[3] = {(a, b, c)}).

It does "work" with `__VA_ARGS__`, - what do you think??